### ISSUE TITLE
Allow explicitly setting classnames for series

### ISSUE THREAD

**Issue by n1313** [chartist-js/chartist#940]
I'm trying to use chartist on my project where I'm using CSS Modules for styling. My classes look like this: `'LineChart--series--f45ba0'`. Note the randomly generated bit at the end. There's a slight problem with this setup: chartist doesn't allow passing in classnames for individual series, and instead attaches the `-a` or `-b` bit to the end of the series classname. In my case, this generates classnames like `'LineChart--series--f45ba0-a'` which is not what I want.

Is this possible? If not, would you accept a PR that adds an option to explicitly pass in classnames for individual series, so that I could do something like
```
    classNames: {
        ...
        series: css.series, // 'LineChart--series--f45ba0'
        'series-a': css.seriesA, // 'LineChart--seriesA--1bac43'
        'series-b': css.seriesB, // 'LineChart--seriesB--eff81c'
        ...
    }
```
?

p.s. is gitter chat alive? I asked this question there almost two weeks ago but nobody picked up the phone.

**Comment by gionkunz** [comment]
Hey there. You can use the object series notation to control the classes of series:

```javascript
new Chartist.Line('.ct-chart', {
  labels: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'],
  series: [{
    data: [12, 9, 7, 8, 5],
    className: 'LineChart--series--f45ba0'
  }, {
    data: [2, 1, 3.5, 7, 3],
    className: 'LineChart--series--f45ba0'
  }, {
    data: [1, 3, 4, 5, 6],
    className: 'LineChart--series--f45ba0'
  }]
});
```

Hope this helps.

**Comment by n1313** [comment]
Yes, this is the approach I'm using, but it's a bit clunky. Would you accept a PR with this change?

**Comment by gionkunz** [comment]
Why is it clunky? :-) Can you describe this in more technical terms and with a clear technical benefit? Also for consistency we would need to deprecate the series object notation class name since we should not have both APIs available. This would be a major change, please provide some technical reasoning why the series data object notation is not sufficient.

**Comment by n1313** [comment]
Certainly.

I am working on integrating chartist into a React component. The component is a part of bigger library of reusable UI components which is used in a large corporate environment. I have to meet these requirements:

1. Minimize the risk of CSS class name collisions. There are many teams and projects in my company, I don't know where my code is going to be used, so I am using CSS Modules and rely on unique auto-generated class names with random bits. And, since there is no guarantee that global `.ct-` namespace is not taken (maybe the project name is CT, or they are already using chartist and have their own customizations in place which I don't want to mess up), I need to take chartist's CSS classes and rename them. So `.ct-series` becomes `.ChartComponent--series--a45bc0` or something like that. This is, thankfully, quite easy to do via chartist configuration and things work fine.

2. Follow the branding. My charts have to be colored in a certain way by defaut: first series is CorporateColor1, second is CorporateColor2, etc.  This could be easily done if I could redefine global styles and assign my colors to `.ct-series-a` and others, but I can't since there is no such class in my stylesheet. Chartist assigns `.ChartComponent--series--a45bc0-a` to the series, and I want it to be `.ChartComponent--seriesA--34bb17`. So I have to iterate over my data and modify each series manually by converting the array into object and then modifying the className, and do this every time my data is changed as well. This works, but compared to the simple configuration change I did in p.1 is what I call "clunky".

I don't see why my proposed change would deprecate anything or be a breaking change. Currently chartist has hidden unchangeable defaults for series class names. I propose to make them changeable via configuration:

```diff
seriesElement.addClass([
  options.classNames.series,
- (series.className || options.classNames.series + '-' + Chartist.alphaNumerate(seriesIndex))
+ (series.className || options.classNames([series + '-' + Chartist.alphaNumerate(seriesIndex)]) || options.classNames.series + '-' + Chartist.alphaNumerate(seriesIndex))
].join(' '));
```
[here](https://github.com/gionkunz/chartist-js/blob/develop/src/scripts/charts/line.js#L167) and then repeat the same for other two chart types. This will allow me to pass in the class names I need and the rest of the world will not notice.

**Comment by gionkunz** [comment]
Therese usecases are already fulfilled by using the series as object notation:

```javascript
new Chartist.Line('.ct-chart', {
  labels: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'],
  series: [{
    data: [12, 9, 7, 8, 5],
    className: 'LineChart--series--f45ba0'
  }, {
    data: [2, 1, 3.5, 7, 3],
    className: 'LineChart--series--f45ba0'
  }, {
    data: [1, 3, 4, 5, 6],
    className: 'LineChart--series--f45ba0'
  }]
});
```

Can you tell me technical reasoning why we should move the class name overrides into options.classNames ?

**Comment by n1313** [comment]
I have provided all the explanation I could in my previous comment. Please read the sample diff I wrote there.

I'm sorry, but I really don't understand the resistance I feel coming from you here. This is a really simple addition I'm proposing. I'm not asking to *move* the declarations, I'm asking to *add* an *optional* way to configure them from the config object. Can you explain what is the problem you're seeing with it?

**Comment by gionkunz** [comment]
Okay, let me try again. 

1. Two solutions for the same problem are always a bad idea. It's confusing, leading to code smell and introduces complexity. If you're provided two ways to tackle the same problem within an API, this is considered a bad thing. If you compare it to science, this would probably be related to https://en.wikipedia.org/wiki/Occam%27s_razor. Therefore I strongly think only one solution should exist. If we're introducing your solution, we would need to remove the current solution. This would make this change a breaking one.

2. Besides that your solution is fixed a problem which is already fixed and we would need to decide on one of both solutions to make it consistent, let's look at both solutions and the pro / cons:

Current solution:
```javascript
new Chartist.Line('.ct-chart', {
  labels: ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday'],
  series: [{
    data: [12, 9, 7, 8, 5],
    className: 'LineChart--series--f45ba0'
  }, {
    data: [2, 1, 3.5, 7, 3],
    className: 'LineChart--series--f45ba0'
  }, {
    data: [1, 3, 4, 5, 6],
    className: 'LineChart--series--f45ba0'
  }]
});
```

You solution:
```javascript
classNames: {
    ...
    series: css.series, // 'LineChart--series--f45ba0'
    'series-a': css.seriesA, // 'LineChart--seriesA--1bac43'
    'series-b': css.seriesB, // 'LineChart--seriesB--eff81c'
    ...
}
```

* Current solution:
  * Pros
    * It's a real override and not a mapping
    * it's coupled to the series in the data so if you reorder your data then the class names will stick to the correct series
  * Cons
    * It may seem the wrong place to have this in your data since all the other class mappings are stored in the `classNames` configuration

* You solution:
  * Pros
    * The class names for individual series are configured in the `classNames` configuration where all other class names can be configured too
  * Cons
    * You're providing a class re-mapping and it's not a direct configuration. I always need to specify the internally generated class name and the new class name. Also the generated class names will change in the future (to be numbered see #183 ) which would then always be a breaking change for users. Actually you're making the configuration of users to be dependent on some internal state (default class names).
    * I may specify too many class mappings, or not enough, and since it's not directly attached to the series, I always need to check back and forth from `classNames` into my series to see if I have all the mappings.

**Conclusion**

I don't see a real benefit in your solution and you did not convince me that it's worth going through a backwards incompatible change just for the sake of your opinion. Also I would never agree to introduce a change that makes some configuration redundant.

I'm not being resistant here, I'm just trying to protect Chartist from complexity and code smell. If I would not do this, the library would quickly suffer from scope creep and inconsistency. If you don't understand this then I'm feeling sorry for you. I'm open for comments, suggestions and you can always convince me of a better solution. But if you can't convince me then I will not merge a PR. There were large PR's I've merged within a few minutes after the PR was made. Because I was immediately convinced that this would be a good thing and no alarm bells were ringing when looking at the overall change of the library. However, there are others, yours included, where I have my doubts.

**Comment by n1313** [comment]
I appreciate your response, it is very thorough.

I'd like to highlight that the current approach and the one I'm proposing are not equal: they are not solving exactly the same problem. Object notation for individual series solves the problem of "I want this particular series to have this particular classname" (which is an override, and not a mapping), when explicitly configuring classnames in instance config solves the problem of "I want every n-th series to have this particular classname" (a global mapping, not an override). These two, of course, can be used to achieve the same result but are not the same thing. The problem I'm facing is that there is *no way* to set a global default for series' styles if I'm using CSS modules or JSS or any other similar library that generates randomised classnames -- without assigning every series a custom classname. Currently, every first series is #d70206, every second is #f05b4f and so on. If I'm using CSS modules and I can't use global classnames, then I can't override these values unless I touch data.

#183 is open for more than two years so forgive me for not understanding that it is a planned change. When is 1.0 planned to be released? Regardless, switching from letters to numbers or doing any other change that changes the current naming scheme is going to be breaking, whether you use the current defaults or use defaults configured with my proposal.

**Comment by n1313** [comment]
...so, @gionkunz, any reply to this?

**Comment by mrksbnch** [comment]
I'm facing a similar issue and being able to set a global default would help.
